### ISSUE TITLE
Mongo Relationships on deep document keys?

### ISSUE THREAD

**Issue by NeoVance** [ikkez/F3-Sugar#42]
Is this possible? If so how is this defined in the fieldConfig property?

**Comment by ikkez** [comment]
hi @NeoVance. Can you be more specific please. I don't quite get what you mean or trying to do.

**Comment by NeoVance** [comment]
A mongo document can be an unlimited number of nested JSON objects and arrays. I would like to create a relationship to keys that are deeper than the first level of the document. For instance: I have a collection of projects each of which contain a collection of tasks, comments, and actions taken. On those sub collections I would like to create a relationship link in Cortex for the sub collections to another main collection where user data is stored.

```
project->comments[0]->user <1==0..*> \Model\User
```

**Comment by ikkez** [comment]
Ahh, now i see. Well there is no such feature yet, but sounds reasonable. I'll think about a proper implementation when i'll find the time. In the meanwhile i just wrote that custom getter for you, that should do the same job as well:

``` php
protected function get_comments($val){
	if ($this->collectionID) {
		$cx = \DB\CortexCollection::instance($this->collectionID);
		if (!$cx->hasRelSet('user')) {
			$users = new User();
			$user_ids=array();
			foreach($cx->getAll('comments',true) as $comments)
				foreach($comments as $comment)
					$user_ids[] = $comment['user'];
			array_unique($user_ids);
			$cx->setRelSet('user',$users->find(array('_id IN ?',$user_ids))->getBy('_id'));
		}
		foreach($val as &$comment) {
			$user_cx = $cx->getSubset('user',$comment['user']);
			$comment['user'] = $user_cx[0]->cast();
			unset($comment);
		}
	}
	return $val;
}
```

just put that function into your project model.

**Comment by ChrisFrench** [comment]
Doctrine does this. It gets to be a configuration nightmare but pretty cool
feature

**Comment by NeoVance** [comment]
The configuration with Cortex should be easier, since it can just use the FatFree hive key syntax with slight modification for keys.

```
// Project
$fieldConf = array( "comments[].user" => array( "belongs-to-one" => "\Model\User" ) );

// User
$fieldConf = array( "comments" => array( "has-many" => array("\Model\Project" => "comment[].user") ) );
```

Eventually, it would be nice if relationships could be defined much like the routes or maps in FatFree.

```
$project->belongsToOne( "comments[].user", "\Model\User" );
$user->hasMany( "comments", "\Model\Project", "comments[].user" );
```

**Comment by ikkez** [comment]
Yeah that are all nice ideas, but the point is that there is no real "table" for comments, if you just nest them in a property of another entity. That makes it hard to the mapper and some relational features. All i can offer now is that loose coupling, one-directional, to inject the user data into your nested array property, but there won't be any Comment model, since there is no table for it. I guess we could emulate that, and add coupling the other way round (`user->comments`) with some group reduce mongo functions running through all projects, but also the `has` and `filter` features can't work as they do now. I mean thinks like "get all my own comments from projects that are in category xy".

**Comment by NeoVance** [comment]
Dynamically replace each sub collection with a CortexCollection hydrated by a new mapper extending the Jig mapper to use a reference to the main models sub collections instead of pulling information from a file? In this case we can simulate an additional table with a relationship of many-to-one with the main table, but not require the additional queries to the database.

Base::cut will come in handy here. We can add the relationship for the sub collection in the field config:
```
$fieldConf = array(
     "comments" => array( 
          "has-many" => array( "\Model\Project->comments", "project" )
          "user" => array( "belongs-to-one" => "\Model\User" )
          "project" => array( "belongs-to-one" => "\Model\Project" )
     )
);
```
When that relationship is lazy loaded it will create a new jig-like database table containing the referenced key's data with the required relationships.

**Comment by NeoVance** [comment]
The changelog for FatFree 3.4 mentions something about provision for in memory Jig database.

**Comment by ikkez** [comment]
yes indeed, an in memory Jig mapper could be used for that. Nice idea. 
I really appreciate your efforts and investigations towards this topic. I guess most people are just sacred when they see its complexity, don't see how it works or just use other solutions,.. so it's nice to see one who makes thoughts about it.
regarding the idea, i could imagine we start here:

```php
// Project
$fieldConf = array( 
  "comments.user" => array( "belongs-to-one" => "\Model\User" )
);
// User
$fieldConf = array(
  "comments" => array( "has-many" => array("\Model\Project" => "comments.user") )
);
```
I think it's okay to remove the extra `[]` in definition,.. should be possible to determine if it's a single array or multiple, based on the `comments` field. So once the field is lazy-loader, or is part of a filter function, a new Memory-Mapper is created that pre-fetches and keeps all the values. This would also be compatible with all other DB engines, which is nice. But on the other hand this way could eat up a lot of memory, and at least for mongo there are ways to fetch and reduce the collection before. But i'm not that deep into mongo yet and have no idea in my mind for that now. 

With the MemoryJig, the only issue now is that we're mixing Mongo with Jig Mappers. Well even if i said that those mash-ups are possible, i must admit i haven't tested those to the end yet, and i should fix that before. I wanted to do that anyways, but i still have some questions in my mind like: When you have a many-many relation from Mongo to Jig, where is the privot-table? should it be in mongo or in jig. I guess that needs to be configurable somehow.

Another configuration-method for the memory-jig type could also be to create a new model class for comments, but instead of defining a table, we point it to the other mapper, like 

```php
// Comment
$table = array('\Model\Project','comments');
$fieldConf = array( 
  "user" => array( "belongs-to-one" => "\Model\User" ),
  "project" => array( "belongs-to-one" => "\Model\Project" ),
);
// Project
$fieldConf = array( 
  "comments" => array( "belongs-to-many" => "\Model\Comment" )
);
// User
$fieldConf = array(
  "comments" => array( "has-many" => array("\Model\Comment" => "user") )
);
```
But i guess there is even more work to do to get that beast running ;)

**Comment by NeoVance** [comment]
I think the memory issue is not something you should have to worry about. The entire document returned by mongo is already stored in memory, we are just creating an object that will be able to access those values (hopefully by reference!) as a type of wrapper.

> When you have a many-many relation from Mongo to Jig, where is the pivot-table? should it be in mongo or in jig.

I think you could make it configurable...by default it should probably be in the Jig, since Mongo is not designed to think about relationships that way, most developers would not create a table like that in Mongo to begin with. I doubt they would want the library tainting their database with this additional information unless it was already planned. At the same time, you will have better performance if you do use the MongoDB so it may be preferable on larger data-sets.

The SQL database could optionally use this behavior as well for fields where the data-type if JSON or SERIALIZED.

Probably the first thing would be to get Cortex updated to work correctly with FatFree 3.4. That release changed the get function in magic. It now returns a reference, and I think there is one less parameter. I think it may have changed something around the getIterator function as well. In any case, I had to make some modification to get it to work without throwing 500 errors, and even then it was having some trouble with "belongs-to-one" relationships. I think there is a loop not exiting correctly, since the script is hitting the time limit.

**Comment by ikkez** [comment]
regarding FatFree 3.4, defenitly check the [dev branch](https://github.com/ikkez/F3-Sugar/tree/dev-cortex/Cortex) It has lot of fixes and is compatible with 3.4. Unfortunately the 3.4 release has some grouping bugs, so the best bet is to use F3 3.4.1-Dev version at https://github.com/bcosca/fatfree-core. I have everything running up and fine with 3.4 at [fabulog](https://github.com/ikkez/fabulog)
I wanted to release Cortex v1.3 with the F3 release, but since there were new bugs in the release, it sadly wouldn't be compatible.

**Comment by NeoVance** [comment]
Been super busy with a project this year, and have not had much time to dig into this much. Now that work has settled down, I was wanting to come back to this. Have you made much progress in this regard as of now?
